src/command/output.rs

```diff
@@ -424,106 +426,43 @@ impl RoverOutput {
         JsonVersion::default()
     }
 
-    pub(crate) fn print_descriptor(&self, descriptor: impl Display) -> io::Result<()> {
+    pub(crate) fn print_descriptor(&self) -> io::Result<()> {
         if atty::is(Stream::Stdout) {
-            stderrln!("{}: \n", Style::Heading.paint(descriptor.to_string()))?;
+            if let Some(descriptor) = self.descriptor() {
+                stderrln!("{}: \n", Style::Heading.paint(descriptor))?;
+            }
         }
         Ok(())
     }
-    pub(crate) fn print_one_line_descriptor(&self, descriptor: impl Display) -> io::Result<()> {
+    pub(crate) fn print_one_line_descriptor(&self) -> io::Result<()> {
         if atty::is(Stream::Stdout) {
-            stderr!("{}: ", Style::Heading.paint(descriptor.to_string()))?;
+            if let Some(descriptor) = self.descriptor() {
+                stderr!("{}: ", Style::Heading.paint(descriptor))?;
+            }
         }
         Ok(())
     }
-}
-
-#[derive(Debug, Clone, Serialize)]
-pub(crate) struct JsonOutput {
-    json_version: JsonVersion,
-    data: JsonData,
-    error: Value,
-}
-
-impl JsonOutput {
-    pub(crate) fn success(data: Value, error: Value, json_version: JsonVersion) -> JsonOutput {
-        JsonOutput {
-            json_version,
-            data: JsonData::success(data),
-            error,
-        }
-    }
-
-    pub(crate) fn failure(data: Value, error: Value, json_version: JsonVersion) -> JsonOutput {
-        JsonOutput {
-            json_version,
-            data: JsonData::failure(data),
-            error,
-        }
-    }
-
-    pub(crate) fn print(&self) -> io::Result<()> {
-        stdoutln!("{}", self)
-    }
-}
-
-impl fmt::Display for JsonOutput {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{}", json!(self))
-    }
-}
-
-impl From<RoverError> for JsonOutput {
-    fn from(error: RoverError) -> Self {
-        let data_json = error.get_internal_data_json();
-        let error_json = error.get_internal_error_json();
-        JsonOutput::failure(data_json, error_json, error.get_json_version())
-    }
-}
-
-impl From<RoverOutput> for JsonOutput {
-    fn from(output: RoverOutput) -> Self {
-        let data = output.get_internal_data_json();
-        let error = output.get_internal_error_json();
-        JsonOutput::success(data, error, output.get_json_version())
-    }
-}
-
-#[derive(Debug, Clone, Serialize)]
-pub(crate) struct JsonData {
-    #[serde(flatten)]
-    inner: Value,
-    success: bool,
-}
-
-impl JsonData {
-    pub(crate) fn success(inner: Value) -> JsonData {
-        JsonData {
-            inner,
-            success: true,
-        }
-    }
-
-    pub(crate) fn failure(inner: Value) -> JsonData {
-        JsonData {
-            inner,
-            success: false,
+    pub(crate) fn descriptor(&self) -> Option<&str> {
```

Comment:
separated this out into its own function and had it return an `Option<&str>` instead of an early `return Ok(())` as it makes more sense semantically